### PR TITLE
Release Links 0.9 (Burghmuirhead).

### PR DESCRIPTION
**Pull request by SimonJF**
0.9 Changelog:

Version 0.9 (Burghmuirhead) presents multiple major new features, bugfixes, as
well as significant changes and improvements to the Links internals.

Links now includes a Model-View-Update library for writing client code, as
pioneered by the Elm programming langauge. More can be found on
[the Wiki page](https://github.com/links-lang/links/wiki/Model-View-Update-(Elm-Architecture)).

The implementation of relational lenses has been extended to support _dynamic_
predicates in `select` queries. Additionally, much work has been done in order
to allow better typechecking of relational lenses.

Mutually-recursive functions and types should now be enclosed in a `mutual`
block:

```
mutual {
  typename Even = [| Z | SuccE:Odd |];
  typename Odd = [| SuccO:Even |];

  sig three : () -> Odd
  fun three() {
    SuccO(SuccE(SuccO(Z)))
  }
}
```

This leads to significant improvements with the performance of the typechecker
when considering mutually-recursive types, simplifies the code, and solves
multiple bugs.

FreezeML is a new approach to integrating first-class, System F-style
polymorphism with ML-style type inference. The system relies on not guessing
polymorphism, and the ability to "freeze" variables in order to suppress
instantiation. See the [TyDe abstract](http://tydeworkshop.org/2019-abstracts/paper19.pdf) for more details.

Version 0.9 includes an implementation of FreezeML:

  - Frozen variables (i.e., variables which are not instantiated) are written
    `~e`
  - Frozen lets (i.e., function definitions which never generalise their types)
    are written `~fun f(x) { x }`
  - Explicit function generalisation is written `$(fun(x) { x }` and produces a
    term of type `forall a::Any,b::Row. (a) -b-> a`
  - Explicit instantiation is written using `@`

The previous work on "explicit quantification" has been removed.

If the `effect_sugar` flag is set, there are many improvements which decrease
the number of times where explicit effect variables are required.

  - Fresh effect variables are no longer generated by the parser
  - Higher-order functions share an effect variable with their calling context: for example,
    `map : ((a) -e-> b, [a]) -e-> [b]` can be written as `map : ((a) -> b, [a]) -> [b]`.
  - Type names can now be written with an implicit effect variable, for example
    `typename Comp(a) = () ~> a` can be written insetad of `Comp(a, e::Eff) = () ~e~> a, and
    `forever : Comp(()) ~> ()` can be written instead of `forever : Comp((), e) ~e~> ()`
  - Presence variables can be omitted from effects. As an example, we can
    write
      ```
      sig evalState : (s) ->
                    (Comp(a, {Get:s,Put:(s) {}-> () |_})) ->
                     Comp(a, {                      |_})
      ```
      instead of
      ```
      sig evalState : (s) ->
                    (Comp(a, {Get:s,Put:(s) {}-> () |_})) ->
                     Comp(a, {Get{_},Put{_}         |_})
      ```

Much work on 0.9 has concentrated on significant non-user-visible changes. In
particular:

  - Desugaring passes which occur after the typechecker are now type-preserving
  - Changes to internal representation of types
  - Many internal refactorings

 - Modules now work better on the REPL

 - `import` is now used to allow a module to be used in the current file,
   whereas `open` brings it into the global scope. To get the previous
   behaviour, use `open import X` for a module `X`. (breaking change)

 - Many behind-the-scenes improvements and bugfixes

 - Compilation on OCaml 4.08 now supported
 - `select` queries now printed when debugging enabled
 - Query shredding now supported in SQLite
 - The `End` session type is now linear, and must be eliminated using `close`.
   This solves a class of memory leaks with session-typed applications.
 - Record field punning is now supported. For example, instead of writing
   `var hello = "hi"; (hello = hello, world="universe")`, it is now possible to
   write `var hello = "hi"; (=hello, world="universe")`.
 - `lines`, `unlines`, `partition`, and `span` added to prelude
 - Hyphenated HTML attributes are now allowed
 - "Primed" variables such as `x'` are now allowed
 - Initial (hacky) support for null integers in the database. If
   the `coerce_null_integers` setting is set to `true`, then `null_integer`
   (default -1) will be used instead of crashing at runtime.
 - Fix draggable list (sessions), draggable cropping frame examples

### CODE REVIEW

File: packages/links-postgresql/links-postgresql.0.9/opam [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+
+build: [
+  [ "dune" "subst" ] {pinned}
+  [ "dune" "build" "-p" name "-j" jobs ]
+]
+
+depends: [
+  "ocaml" {>= "4.06.0"}
+  "dune" {build}
```
Comment (kit-ty-kate) [inline packages/links-postgresql/links-postgresql.0.9/opam]:
```suggestion
  "dune" {>= "1.10"}
```
according to the `dune-project` file, this project requires dune >= 1.10. Also, dune should not be tagged as a build dependency (see #14266)

File: packages/links-sqlite3/links-sqlite3.0.9/opam [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+license: "GPL-2"
+
+build: [
+  [ "dune" "subst" ] {pinned}
+  [ "dune" "build" "-p" name "-j" jobs ]
+]
+
+depends: [
+  "ocaml" {>= "4.06.0"}
+  "dune" {build}
```
Comment (kit-ty-kate) [inline packages/links-sqlite3/links-sqlite3.0.9/opam]:
```suggestion
  "dune" {>= "1.10"}
```

File: packages/links/links.0.9/opam [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+build: [
+  [ "dune" "subst" ] {pinned}
+  [ "dune" "exec" "preinstall/preinstall.exe" "--" "-libdir" _:lib ]
+  [ "dune" "build" "-p" name "-j" jobs ]
+]
+
+depends: [
+  "ocaml" {>= "4.06.0"}
+  "dune" {build}
```
Comment (kit-ty-kate) [inline packages/links/links.0.9/opam]:
```suggestion
  "dune" {>= "1.10"}
```

**Comment by SimonJF** [comment]
Thanks, Kate, I appreciate the review and suggestions! I believe @dhil has fixed these issues upstream, too, so we should be future-proofed in that regard.

**Comment by mseri** [comment]
Thanks

**Pull request merged**